### PR TITLE
EZP-30693: Refactored SolrEngineFactory to avoid using container directly

### PR DESCRIPTION
**Pull request by webhdx**
> JIRA: https://jira.ez.no/browse/EZP-30693

Introduced registries and service tagging to allow fetching Gateways and CoreFilters for given connection. This avoid direct `Container::get` calls and let us keep services private.

### CODE REVIEW

File: lib/Container/Compiler/CoreFilterRegistryPass.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+class CoreFilterRegistryPass implements CompilerPassInterface
+{
+    /**
+     * @param \Symfony\Component\DependencyInjection\ContainerBuilder $container
+     *
+     * @throws \LogicException
+     */
+    public function process(ContainerBuilder $container)
+    {
+        if (!$container->hasDefinition('ezpublish.search.solr.core_filter.registry')) {
```
Comment (adamwojs) [inline lib/Container/Compiler/CoreFilterRegistryPass.php]: Could you please extract constant with registry service id?

File: lib/Container/Compiler/CoreFilterRegistryPass.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+            return;
+        }
+
+        $coreFilterRegistryDefinition = $container->getDefinition('ezpublish.search.solr.core_filter.registry');
+
+        $coreFilters = $container->findTaggedServiceIds('ezpublish.search.solr.core_filter');
```
Comment (adamwojs) [inline lib/Container/Compiler/CoreFilterRegistryPass.php]: Could you please extract constant with tag name?

File: lib/Container/Compiler/GatewayRegistryPass.php [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+class GatewayRegistryPass implements CompilerPassInterface
+{
+    /**
+     * @param \Symfony\Component\DependencyInjection\ContainerBuilder $container
+     *
+     * @throws \LogicException
+     */
+    public function process(ContainerBuilder $container)
+    {
+        if (!$container->hasDefinition('ezpublish.search.solr.gateway.registry')) {
```
Comment (adamwojs) [inline lib/Container/Compiler/GatewayRegistryPass.php]: Could you please extract constant with gateway service id?

File: lib/Container/Compiler/GatewayRegistryPass.php [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+            return;
+        }
+
+        $gatewayRegistryDefinition = $container->getDefinition('ezpublish.search.solr.gateway.registry');
+
+        $gateways = $container->findTaggedServiceIds('ezpublish.search.solr.gateway');
```
Comment (adamwojs) [inline lib/Container/Compiler/GatewayRegistryPass.php]: Could you please extract constant with tag name?

File: lib/Container/Compiler/CoreFilterRegistryPass.php [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+ * @version //autogentag//
+ */
+namespace EzSystems\EzPlatformSolrSearchEngine\Container\Compiler;
+
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Reference;
+use LogicException;
+
+class CoreFilterRegistryPass implements CompilerPassInterface
```
Comment (adamwojs) [inline lib/Container/Compiler/CoreFilterRegistryPass.php]:
```suggestion
final class CoreFilterRegistryPass implements CompilerPassInterface
```

File: lib/Container/Compiler/GatewayRegistryPass.php [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+ * @version //autogentag//
+ */
+namespace EzSystems\EzPlatformSolrSearchEngine\Container\Compiler;
+
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Reference;
+use LogicException;
+
+class GatewayRegistryPass implements CompilerPassInterface
```
Comment (adamwojs) [inline lib/Container/Compiler/GatewayRegistryPass.php]:
```suggestion
final class GatewayRegistryPass implements CompilerPassInterface
```

File: lib/CoreFilter/CoreFilterRegistry.php [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ *
+ * @version //autogentag//
+ */
+namespace EzSystems\EzPlatformSolrSearchEngine\CoreFilter;
+
+use EzSystems\EzPlatformSolrSearchEngine\CoreFilter;
+use OutOfBoundsException;
+
+class CoreFilterRegistry
```
Comment (adamwojs) [inline lib/CoreFilter/CoreFilterRegistry.php]:
```suggestion
final class CoreFilterRegistry
```

File: lib/Gateway/GatewayRegistry.php [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+ */
+namespace EzSystems\EzPlatformSolrSearchEngine\Gateway;
+
+use EzSystems\EzPlatformSolrSearchEngine\Gateway;
+use OutOfBoundsException;
+
+/**
+ * Registry for Solr search engine coreFilters.
+ */
+class GatewayRegistry
```
Comment (adamwojs) [inline lib/Gateway/GatewayRegistry.php]:
```suggestion
final class GatewayRegistry
```

**Comment by webhdx** [comment]
@adamwojs We are still looking for a volunteer. :)

**Comment by Nattfarinn** [comment]
We're run out of juniors and interns :P

**Comment by webhdx** [comment]
@adamwojs What exactly do you want to test? I'm fine with Compiler Passes but there is little sense in testing Registries as they only contain getters and setters.

**Comment by adamwojs** [comment]
> @adamwojs What exactly do you want to test? I'm fine with Compiler Passes but there is little sense in testing Registries as they only contain getters and setters.

Tests for Compiler Passes should be enough. Personally, I think registries are also worth to cover in tests: there are still services, even if it's logic is very simple. But this is my personal opinion.  

BTW.  Failed CI build seems to be relevant 😉

**Comment by webhdx** [comment]
@adamwojs Added tests for Registries and CompilerPasses. Travis should be green now.

**Comment by adamwojs** [comment]
> @adamwojs Added tests for Registries and CompilerPasses. Travis should be green now.

Thank you @webhdx 🍻

**Pull request merged**